interp: correctly mark functions as modifying memory

aykevl opened the pull request:
This is a fix for an interp bug I found in #4732.
Making a draft PR since I want to see the impact of this, it could be big. I might need to modify the PR.

aykevl committed:
interp: correctly mark functions as modifying memory

Previously, if a function couldn't be interpreted in the interp pass, it
would just be run at runtime and the parameters would be marked as
potentially being modified. However, this is incorrect: the function
itself can also modify memory. So the function itself also needs to be
marked (recursively).

This fixes a number of regressions while upgrading to Go 1.24.

This results in a significant size regression that is mostly worked
around in the next commit.

deadprogram commented:
```
--- FAIL: TestBinarySize (0.00s)
    --- FAIL: TestBinarySize/microbit/examples/serial (40.40s)
        sizes_test.go:67: Unexpected code size when compiling: -target=microbit examples/serial
        sizes_test.go:68:             code rodata   data    bss
        sizes_test.go:69: expected:   2908    388      8   [22](https://github.com/tinygo-org/tinygo/actions/runs/13391197573/job/37399093533?pr=4742#step:15:23)72
        sizes_test.go:70: actual:     4352    404    176   2552
    --- FAIL: TestBinarySize/wioterminal/examples/pininterrupt (41.07s)
        sizes_test.go:67: Unexpected code size when compiling: -target=wioterminal examples/pininterrupt
        sizes_test.go:68:             code rodata   data    bss
        sizes_test.go:69: expected:   7293   1487    116   6912
        sizes_test.go:70: actual:     8691   1501    [28](https://github.com/tinygo-org/tinygo/actions/runs/13391197573/job/37399093533?pr=4742#step:15:29)8   7200
FAIL
```

Impact was not small on your first attempt, it would seem. :elephant:

aykevl commented:
Oops, yeah, I was afraid of this.
I'll need to investigate this a bit deeper.

aykevl committed:
runtime: manually initialize xorshift state

This ensures:

 1. The xorshift state is initialized during interp.
 2. The xorshift state gets initialized to a real random number on
    hardware that supports it at runtime.

This fixes a big binary size regression from the previous commit. It's
still not perfect: most programs increase binary size by a few bytes.
But it's not nearly as bad as before.

aykevl commented:
Trying a workaround that should hopefully fix the worst of this. Let's see what happens.

aykevl commented:
So this is fixing a real bug, but also uncovering a bunch of cases where the old behavior was incorrect but resulted in smaller binary sizes. Most importantly (at least for small programs), the `runtime.fastrand` function was previously just returning 0 during interp since `xorshift32State` wasn't initialized but interp (incorrectly) continued interpreting it anyway. This bugfix fixes this issue - but also means that callers of fastrand (most notably the hashmap code) now can't be run during interp.

I've added a fix that should fix the worst of this, by initializing the xorshift state to a valid (non-random) value so that fastrand actually returns pseudo-random numbers that can be used by interp while still making fastrand use hardware random numbers at runtime.

@dgryski you wrote some of this, what do you think of this solution?

deadprogram commented:
@aykevl I think we can afford the extra 2 bytes here https://github.com/tinygo-org/tinygo/actions/runs/13415479905/job/37475187142?pr=4742#step:19:16 :smiley_cat:

aykevl commented:
Oops, yeah I missed that one.

deadprogram commented:
@aykevl looking at the sizediff, it would appear that this PR is a success.

dgryski commented:
I have the feeling this will fix a number of other interp-related issues on our bug tracker.

deadprogram reviewed (approved):
Thanks for working on this @aykevl

deadprogram commented:
I will merge after the current build finishes.

deadprogram commented:
Now merging, thanks @aykevl

dgryski commented:
I tracked down a regression that was preventing an internal service from building with recent releases of tinygo to this.  This PR causes the `asyncify` pass to take 10x longer for a 10% binary increase size, and the `precompute-propagate` to basically never finish.
